### ISSUE TITLE
cloud-tests tree_run/bddeb fail to build deb

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#3362]
This bug was originally filed in Launchpad as [LP: #1822615](https://bugs.launchpad.net/cloud-init/+bug/1822615)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = []
assignee = None
assignee_name = None
date_closed = None
date_created = 2019-04-01T14:15:51.647327+00:00
date_fix_committed = None
date_fix_released = None
id = 1822615
importance = low
is_complete = False
lp_url = https://bugs.launchpad.net/cloud-init/+bug/1822615
milestone = None
owner = raharper
owner_name = Ryan Harper
private = False
status = triaged
submitter = raharper
submitter_name = Ryan Harper
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **Ryan Harper(raharper)** wrote on 2019-04-01T14:15:51.647327+00:00_

While looking to verify an apt_pipelining test-case fix I ran this:

% tox -e citest -- tree_run --verbose --os-name bionic --test modules/apt_pipelining_disable --test modules/apt_pipelining_os --result /tmp/result.yaml

During which, it would fail like this:

2019-03-31 16:13:11,743 - tests.cloud_tests - ERROR - stage part: build deb on system encountered error: Unexpected error while running command.
Command: /root/packages/bddeb -d
Exit code: 1
Reason: -

....

 ======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_dumps_all_instance_data
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 153, in test_handle_args_dumps_all_instance_data
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_list_keys_errors_when_varname_is_not_a_dict
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 255, in test_handle_args_list_keys_errors_when_varname_is_not_a_dict
    self.assertEqual(1, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_list_keys_sorts_nested_keys_when_varname
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 239, in test_handle_args_list_keys_sorts_nested_keys_when_varname
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_list_keys_sorts_top_level_keys_when_no_varname
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 224, in test_handle_args_list_keys_sorts_top_level_keys_when_no_varname
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 

======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_returns_nested_varname
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 180, in test_handle_args_returns_nested_varname
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_returns_standardized_vars_to_top_level_aliases
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 209, in test_handle_args_returns_standardized_vars_to_top_level_aliases
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
======================================================================       
ERROR: cloudinit.cmd.tests.test_query.TestQuery.test_handle_args_returns_top_level_varname
----------------------------------------------------------------------       
Traceback (most recent call last):                                           
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/tests/test_query.py", line 168, in test_handle_args_returns_top_level_varname
    self.assertEqual(0, query.handle_args('anyname', args))                  
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/cmd/query.py", line 124, in handle_args
    instance_data['userdata'] = util.load_file(user_data_fn)                 
  File "/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412/cloudinit/util.py", line 1359, in load_file
    with open(fname, 'rb') as ifh:                                           
FileNotFoundError: [Errno 2] No such file or directory: 'ud'                 
                                                                             
----------------------------------------------------------------------       
Ran 1945 tests in 16.445s                                                    
                                                                             
FAILED (SKIP=12, errors=7)                                                   
Makefile:54: recipe for target 'unittest3' failed                            
make[2]: *** [unittest3] Error 1                                             
make[2]: Leaving directory '/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412'
debian/rules:21: recipe for target 'override_dh_auto_test' failed            
make[1]: *** [override_dh_auto_test] Error 2                                 
make[1]: Leaving directory '/run/cloud-init/tmp/tmp2v5cncbz/cloud-init-18.5-57-g6c258412'
debian/rules:8: recipe for target 'build' failed                             
make: *** [build] Error 2                                                    
dpkg-buildpackage: error: debian/rules build gave error exit status 2        
                Stderr: debuild: fatal error at line 1376:                   
                        dpkg-buildpackage -rfakeroot -d -us -uc failed

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Ryan Harper(raharper)** wrote on 2019-04-01T14:23:23.367328+00:00_

I worked around this by

1) packaging the tree myself with:

% package/bddeb

2) running citest like this:

tox -e citest -- run --verbose --os-name bionic --test modules/apt_pipelining_disable --test modules/apt_pipelining_os --result /tmp/result.yaml --deb cloud-init_all.deb

----------------------------------------------------------------------
Ran 6 tests in 0.002s

OK (skipped=2)
2019-04-01 13:58:36,162 - tests.cloud_tests - DEBUG - verifying test data for modules/apt_pipelining_os
test_instance_data_json_ec2 (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
Validate instance-data.json content by ec2 platform. ... skipped 'Skipping ec2 instance-data.json on lxd'
test_instance_data_json_kvm (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
Validate instance-data.json content by nocloud-kvm platform. ... skipped 'Skipping nocloud-kvm instance-data.json on lxd'
test_instance_data_json_lxd (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
Validate instance-data.json content by lxd platform. ... ok
test_no_stages_errors (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
Ensure that there were no errors in any stage. ... ok
test_no_warnings_in_log (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
Unexpected warnings should not be found in the log. ... ok
test_os_pipelining (tests.cloud_tests.testcases.modules.apt_pipelining_os.TestAptPipeliningOS)
test 'os' settings does not write apt config file. ... ok

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Paride Legovini(paride)** wrote on 2019-05-10T13:25:50.257340+00:00_

I've been hit by this and investigated a bit. The problem is related to the fact that debuild (and hence the unit tests) are run as root in the container which tree_run spawns to build the .deb package. I can reproduce the problem for example by jumping in the container and running packages/bddeb manually in /root (where the source tree is placed). The exact same command succeeds if I copy the source tree in /home/ubuntu, chmod it ubuntu:ubuntu and run it from there.

I couldn't find where the problem lies exactly. There are some SELinux calls which may affect the tests; I tried to disable SELinux (by modifying the SeLinuxGuard class in util.py), but the problem didn't go away. Another difference is the tmp directory which is used for the tests. Regular users use /tmp, while root uses /run/cloud-init/tmp (see cloudinit/temp_utils.py). As a quick test I made root use /tmp too, but it didn't help.

Having tree_run build the package in the container as a the 'ubuntu' user would probably solve the issue at the cost of adding unneeded complexity; unless the test is failing as root for a good reason I don't favor this idea.
